### PR TITLE
GDT: upload URL request steps wrapped into promises

### PR DESCRIPTION
**Pull request by maksymmalyhin**
Next (close to final) step of the promise refactoring b/160018519.

- GoogleUtilities/Environment dependency introduced to avoid additional code duplication
- URL request logic broken down to steps wrapped into promises
- next upload time logic recovered
- integration tests fixes

#no-changelog

### CODE REVIEW

File: GoogleDataTransport/GDTCCTLibrary/GDTCCTUploadOperation.m [hunk 1]
```diff
@@ -185,99 +183,81 @@ - (void)uploadTarget:(GDTCORTarget)target withConditions:(GDTCORUploadConditions
...
-                     // were uploaded. Useful for validation during tests.
-                     [[NSNotificationCenter defaultCenter]
-                         postNotificationName:GDTCCTUploadCompleteNotification
-                                       object:@(events.count)];
-#endif  // #if !NDEBUG
-                     self.isCurrentlyUploading = NO;
-                     completion();
-                   });
-                 }];
+/** Sends URL request for to upload the provided batch and handles the response. */
```
Comment (paulb777) [inline GoogleDataTransport/GDTCCTLibrary/GDTCCTUploadOperation.m:186]: s/for//
s/handles/handle/

File: GoogleDataTransport/GDTCCTLibrary/GDTCCTUploadOperation.m [hunk 2]
```diff
@@ -185,99 +183,81 @@ - (void)uploadTarget:(GDTCORTarget)target withConditions:(GDTCORUploadConditions
...
+          self.uploaderQueue,
+          ^FBLPromise<NSNull *> *(GULURLSessionDataResponse *response) {
+            // 2. Parse response and update the next upload time if can.
+            [self updateNextUploadTimeWithResponse:response forTarget:target];
+
+            // 3. Cleanup batch.
+
+            // Only retry if one of these codes is returned.
+            NSInteger statusCode = response.HTTPResponse.statusCode;
+            if (statusCode == 429 || statusCode == 503) {
```
Comment (paulb777) [inline GoogleDataTransport/GDTCCTLibrary/GDTCCTUploadOperation.m:204]: Please add comment about 429 and 503 definition

**Review by paulb777 — APPROVED** [review]

File: GoogleDataTransport/GDTCCTLibrary/GDTCCTUploadOperation.m (hunk 2 above)
Comment (maksymmalyhin) [reply GoogleDataTransport/GDTCCTLibrary/GDTCCTUploadOperation.m:206]: added

File: GoogleDataTransport/GDTCCTLibrary/GDTCCTUploadOperation.m (hunk 1 above)
Comment (maksymmalyhin) [reply GoogleDataTransport/GDTCCTLibrary/GDTCCTUploadOperation.m]: done

**Comment by maksymmalyhin** [comment]
I'm going to merge the PR soon, but feel free to comment afterwards.

**Pull request merged**